numba/types/abstract.py

```diff
@@ -209,9 +249,213 @@ class Hashable(Type):
     """
     Base class for hashable types.
     """
+    @classmethod
+    def interface_check(cls, instance):
```

Comment:
I imagine interface checking machinery will allow more flexible typing logic.  A `isinstance(x, Hashable)` is enough for testing whether a type support hashing.  Since `set` is defined to accept `Hashable`, we can have sets of hashable jitclasses as soon as we support nesting memory-managed objects.